File: ZhuZhiLiao/Earth/EarthFeatureModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class EarthFeatureModel {
    enum LoadState: Equatable {
        case loading
        case loaded
        case failed(String)
    }

    private let coordinator: ExperienceCoordinator
    private let locationService: EarthLocationService

    private(set) var loadState: LoadState = .loading
    private(set) var nodes: [EarthNode] = []
    private(set) var serverClockOffsetMilliseconds: Int64 = 0
    private(set) var detail = 2
    private(set) var isUpdatingLocation = false
    private(set) var isLeavingEarth = false
    var selectedNode: EarthNode?
    var showsJoinExplanation = false
    var showsLeaveConfirmation = false

    @ObservationIgnored private var refreshTask: Task<Void, Never>?
    @ObservationIgnored private var detailTask: Task<Void, Never>?
    @ObservationIgnored private var audioExpiryTask: Task<Void, Never>?
    @ObservationIgnored private var isAudioActive = false

    init(
        coordinator: ExperienceCoordinator,
        locationService: EarthLocationService = EarthLocationService()
    ) {
        self.coordinator = coordinator
        self.locationService = locationService
    }

    var isParticipating: Bool { coordinator.earthIsEnabled }
    func start() async {
        await refresh(showLoading: true)
        if isParticipating, locationService.canRefreshWithoutPrompt {
            await refreshExistingLocation()
        }
        synchronizeAudio()
    }

    func startAudio() {
        isAudioActive = true
        synchronizeAudio()
    }

    func stopAudio() {
        isAudioActive = false
        audioExpiryTask?.cancel()
        audioExpiryTask = nil
    }

    func synchronizeAudio() {
        audioExpiryTask?.cancel()
        guard isAudioActive else {
            audioExpiryTask = nil
            return
        }
        guard let nextExpiry = coordinator.synchronizeEarthAudio(
            nodes: nodes,
            serverClockOffsetMilliseconds: serverClockOffsetMilliseconds
        ) else {
            audioExpiryTask = nil
            return
        }

        let delay = max(nextExpiry.timeIntervalSinceNow + 0.02, 0.02)
        audioExpiryTask = Task { [weak self] in
            try? await Task.sleep(for: .seconds(delay))
            guard let self, !Task.isCancelled else { return }
            self.synchronizeAudio()
        }
    }

    func refreshForRevision() {
        refreshTask?.cancel()
        refreshTask = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(120))
            guard let self, !Task.isCancelled else { return }
            await self.refresh(showLoading: false)
        }
    }

    func retry() {
        refreshTask?.cancel()
        refreshTask = Task { [weak self] in
            await self?.refresh(showLoading: self?.nodes.isEmpty == true)
        }
    }

    func setDetail(_ value: Int) {
        let clamped = min(max(value, 0), 4)
        guard clamped != detail else { return }
        detail = clamped
        detailTask?.cancel()
        detailTask = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(180))
            guard let self, !Task.isCancelled else { return }
            await self.refresh(showLoading: false)
        }
    }

    func joinWithCurrentLocation() async {
        isUpdatingLocation = true
        defer { isUpdatingLocation = false }
        do {
            let location = try await locationService.requestOneLocation()
            guard let cellID = EarthLocationGrid.cellID(
                latitude: location.coordinate.latitude,
                longitude: location.coordinate.longitude
            ) else {
                throw EarthLocationError.invalidLocation
            }
            if coordinator.earthCellID != cellID || !coordinator.earthIsEnabled {
                try await coordinator.setEarthLocation(cellID: cellID)
            }
            showsJoinExplanation = false
            await refresh(showLoading: false)
        } catch is CancellationError {
            return
        } catch {
            loadState = .failed(error.localizedDescription)
        }
    }

    func leaveEarth() async {
        isLeavingEarth = true
        defer { isLeavingEarth = false }
        do {
            try await coordinator.disableEarth()
            showsLeaveConfirmation = false
            selectedNode = nil
            await refresh(showLoading: false)
        } catch {
            loadState = .failed(error.localizedDescription)
        }
    }

    private func refresh(showLoading: Bool) async {
        if showLoading, nodes.isEmpty {
            loadState = .loading
        }
        do {
            let snapshot = try await coordinator.loadEarthSnapshot(detail: detail)
            guard !Task.isCancelled else { return }
            let localNow = Int64(Date().timeIntervalSince1970 * 1_000)
            serverClockOffsetMilliseconds = snapshot.serverTime - localNow
            nodes = snapshot.nodes
            if let selectedNode {
                self.selectedNode = nodes.first(where: { $0.id == selectedNode.id })
            }
            loadState = .loaded
            synchronizeAudio()
        } catch is CancellationError {
            return
        } catch {
            loadState = .failed(error.localizedDescription)
        }
    }

    private func refreshExistingLocation() async {
        do {
            let location = try await locationService.requestOneLocation()
            guard let cellID = EarthLocationGrid.cellID(
                latitude: location.coordinate.latitude,
                longitude: location.coordinate.longitude
            ), cellID != coordinator.earthCellID else { return }
            try await coordinator.setEarthLocation(cellID: cellID)
            await refresh(showLoading: false)
        } catch {
            // A stale location is preferable to replacing a successfully loaded
            // globe with an error when an opportunistic refresh fails.
        }
    }
}

File: android/app/src/main/java/com/azhegezhege/zhuzhiliao/earth/EarthFeatureModel.kt
package com.azhegezhege.zhuzhiliao.earth

import com.azhegezhege.zhuzhiliao.ExperienceCoordinator
import com.azhegezhege.zhuzhiliao.network.EarthNode
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

data class EarthFeatureState(
    val isLoading: Boolean = true,
    val error: String? = null,
    val nodes: List<EarthNode> = emptyList(),
    val serverClockOffsetMilliseconds: Long = 0L,
    val selectedNode: EarthNode? = null,
    val isParticipating: Boolean = false,
    val isUpdatingLocation: Boolean = false,
    val isLeaving: Boolean = false,
)

class EarthFeatureModel(
    private val coordinator: ExperienceCoordinator,
    private val scope: CoroutineScope,
    private val onChange: (EarthFeatureState) -> Unit,
) {
    var state = EarthFeatureState(isParticipating = coordinator.uiState.earthIsEnabled)
        private set
    private var detail = 2
    private var refreshJob: Job? = null

    fun start() {
        refresh(showLoading = true)
    }

    fun refreshForRevision() {
        scheduleRefresh(delayMilliseconds = 120, showLoading = false)
    }

    fun retry() = refresh(showLoading = state.nodes.isEmpty())

    fun setDetail(value: Int) {
        val clamped = value.coerceIn(0, 4)
        if (clamped == detail) return
        detail = clamped
        scheduleRefresh(delayMilliseconds = 180, showLoading = false)
    }

    fun select(node: EarthNode?) = update(state.copy(selectedNode = node))

    fun join(locationService: EarthLocationService, after: (Boolean) -> Unit = {}) {
        if (state.isUpdatingLocation) return
        scope.launch {
            update(state.copy(isUpdatingLocation = true, error = null))
            val succeeded = try {
                val location = locationService.requestOneLocation()
                val cellID = EarthLocationGrid.cellID(location.latitude, location.longitude)
                    ?: throw EarthLocationException("当前位置无效，请稍后再试")
                if (coordinator.uiState.earthCellID != cellID || !state.isParticipating) {
                    coordinator.setEarthLocation(cellID)
                }
                update(state.copy(isParticipating = true))
                refresh(showLoading = false)
                true
            } catch (_: CancellationException) {
                false
            } catch (error: Throwable) {
                update(state.copy(error = error.localizedMessage ?: "当前位置更新失败"))
                false
            } finally {
                update(state.copy(isUpdatingLocation = false))
            }
            after(succeeded)
        }
    }

    fun refreshExistingLocation(locationService: EarthLocationService) {
        if (!state.isParticipating || !locationService.canRefreshWithoutPrompt) return
        scope.launch {
            runCatching { locationService.requestOneLocation() }.getOrNull()?.let { location ->
                val cellID = EarthLocationGrid.cellID(location.latitude, location.longitude) ?: return@let
                if (cellID != coordinator.uiState.earthCellID) {
                    runCatching { coordinator.setEarthLocation(cellID) }
                    refresh(showLoading = false)
                }
            }
        }
    }

    fun leave(after: (Boolean) -> Unit = {}) {
        if (state.isLeaving) return
        scope.launch {
            update(state.copy(isLeaving = true, error = null))
            val succeeded = runCatching { coordinator.disableEarth() }
                .onSuccess {
                    update(state.copy(isParticipating = false, selectedNode = null))
                    refresh(showLoading = false)
                }
                .onFailure { update(state.copy(error = it.localizedMessage ?: "退出失败，请稍后重试")) }
                .isSuccess
            update(state.copy(isLeaving = false))
            after(succeeded)
        }
    }

    fun showError(message: String) = update(state.copy(error = message, isLoading = false))

    fun close() {
        refreshJob?.cancel()
    }

    private fun refresh(showLoading: Boolean) {
        scheduleRefresh(delayMilliseconds = 0, showLoading = showLoading)
    }

    private fun scheduleRefresh(delayMilliseconds: Long, showLoading: Boolean) {
        refreshJob?.cancel()
        refreshJob = scope.launch {
            if (delayMilliseconds > 0) delay(delayMilliseconds)
            refreshSuspend(showLoading)
        }
    }

    private suspend fun refreshSuspend(showLoading: Boolean) {
        if (showLoading && state.nodes.isEmpty()) update(state.copy(isLoading = true, error = null))
        try {
            val snapshot = coordinator.loadEarthSnapshot(detail)
            val selectedID = state.selectedNode?.id
            update(
                state.copy(
                    isLoading = false,
                    error = null,
                    nodes = snapshot.nodes,
                    serverClockOffsetMilliseconds = snapshot.serverTime - System.currentTimeMillis(),
                    selectedNode = selectedID?.let { id -> snapshot.nodes.firstOrNull { it.id == id } },
                    isParticipating = state.isParticipating,
                ),
            )
        } catch (_: CancellationException) {
            return
        } catch (error: Throwable) {
            update(state.copy(isLoading = false, error = error.localizedMessage ?: "暂时无法载入哇声地球"))
        }
    }

    private fun update(value: EarthFeatureState) {
        state = value
        onChange(value)
    }
}
